system-contracts/contracts/precompiles/EcPairing.yul

```diff
@@ -0,0 +1,1743 @@
+            /// @dev See https://en.wikipedia.org/wiki/Montgomery_modular_multiplication#The_REDC_algorithm for further details on the Montgomery multiplication.
+            /// @param multiplicand The multiplicand in Montgomery form.
+            /// @param multiplier The multiplier in Montgomery form.
+            /// @return ret The result of the Montgomery multiplication.
+            function montgomeryMul(multiplicand, multiplier) -> ret {
+                let higher_half_of_product := getHighestHalfOfMultiplication(multiplicand, multiplier)
+                let lowest_half_of_product := mul(multiplicand, multiplier)
+                ret := REDC(lowest_half_of_product, higher_half_of_product)
+            }
+
+            /// @notice Computes the Montgomery modular inverse skipping the Montgomery reduction step.
+            /// @dev The Montgomery reduction step is skipped because a modification in the binary extended Euclidean algorithm is used to compute the modular inverse.
+            /// @dev See the function `binaryExtendedEuclideanAlgorithm` for further details.
+            /// @param a The field element in Montgomery form to compute the modular inverse of.
+            /// @return invmod The result of the Montgomery modular inverse (in Montgomery form).
+            function montgomeryModularInverse(a) -> invmod {
+                invmod := binaryExtendedEuclideanAlgorithm(a)
+            }
+
+			// CURVE ARITHMETIC
+
+            /// @notice Checks if a coordinate is on the curve group order.
+            /// @dev A coordinate is on the curve group order if it is on the range [0, curveFieldOrder).
+            /// @param coordinate The coordinate to check.
+            /// @return ret True if the coordinate is in the range, false otherwise.
+            function coordinateIsOnFieldOrder(coordinate) -> ret {
+                ret := lt(coordinate, P())
+            }
+
+            // G1
+
+            /// @notice Checks if a point of the G1 curve is infinity.
+            /// @dev In affine coordinates the infinity is represented by the point (0,0).
+            /// @param x The x coordinate to check.
+            /// @param y The y coordinate to check.
+            /// @return ret True if the point is infinity, false otherwise.
+            function g1AffinePointIsInfinity(x, y) -> ret {
+                ret := iszero(or(x, y))
+            }
+
+            /// @notice Checks if a point in affine coordinates in Montgomery form is on the curve.
+            /// @dev The curve in question is the alt_bn128 curve.
+            /// @dev The Short Weierstrass equation of the curve is y^2 = x^3 + 3.
+            /// @param x The x coordinate of the point in Montgomery form.
+            /// @param y The y coordinate of the point in Montgomery form.
+            /// @return ret True if the point is on the curve, false otherwise.
+			function g1AffinePointIsOnCurve(x, y) -> ret {
```

Comment:
probably mix of tabs/spaces, it looks weird in the github PR diff
<img width="717" alt="image" src="https://github.com/matter-labs/era-contracts/assets/52464764/ce3f92d5-4c19-488d-a6b6-7a75dd385688">

Reply:
Fixed in this commit [99af881](https://github.com/matter-labs/era-contracts/pull/383/commits/99af881ddda1886b7f374ddd17c3bc725233aa39)